### ISSUE TITLE
Type narrowing only works with isinstance()?

### ISSUE THREAD

**Issue by karolyi** [microsoft/pyright#3797]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Hey,

as mentioned [in here](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md?plain=1#L171), type narrowing should work with `type(x) == something`. Maybe I'm missing something but it does not for me.

**To Reproduce**
```python
from datetime import date
from typing import Union

def test(x: Union[str, date]) -> str:
    if type(x) is date:
        return x.isoformat()
    return x
```

**Expected behavior**
Type narrowing should recognize that on the last line, x is a string type.

**Screenshots or Code**
Pointing on `return x`
![image](https://user-images.githubusercontent.com/987055/183247154-f56538be-c21a-4d21-bf0d-86ba370af18c.png)

**VS Code extension or command-line**
I use Sublime-LSP with pyright that updates itself automatically to the latest version.

**Additional context**
I might be missing something but imo it should work with `type(x) is date` as well.

**Comment by erictraut** [comment]
Pyright is doing the right thing here.

Type narrowing is applied in the positive ("if") case, but it is not safe to apply in the negative ("else") case here. That's because the parameter `x` is declared to be type `Union[str, date]` which means it derives from `str` or `date`. It doesn't mean that it has to be `str` or `date`. It could be a class derived from `date`.

As further evidence that pyright is doing the right thing here, mypy's type narrowing behavior is the same.

I recommend switching to an `isinstance` check, which will handle both the positive and negative type narrowing case.

**Comment by karolyi** [comment]
thanks.

in that case,wouldn't the above linked documentation need updating?

**Comment by erictraut** [comment]
The documentation indicates that type narrowing is supported for the pattern `type(x) is T` and `type(x) is not T`, and it is indeed supported.

Each type narrowing pattern contains its own logic that determines how type narrowing is applied in both the positive ("if") and negative ("else") case. For each pattern, there are limits to what can be done safely in each of these cases.

This particular pattern isn't safe to apply in the negative case, at least not in the way you are using it. Narrowing can be applied in the negative case if the type is marked `@final`, but `date` is not marked `@final`, so it can be subclassed.

**Comment by karolyi** [comment]
Thanks you for the elaborate answer! Understood.
